File: Checksum/C++/receiver.cpp
#include <iostream>
#include <string>
#include <bitset>
#include <cstdint>

uint32_t calculateCRC32(const std::string& binaryMessage) {
    uint32_t crc = 0xFFFFFFFF;
    uint32_t crcTable[256];
    uint32_t polynomial = 0xEDB88320;

    // Generate CRC-32 table
    for (uint32_t i = 0; i < 256; ++i) {
        crc = i;
        for (uint32_t j = 0; j < 8; ++j) {
            if (crc & 1) {
                crc = (crc >> 1) ^ polynomial;
            } else {
                crc >>= 1;
            }
        }
        crcTable[i] = crc;
    }

    // Calculate CRC-32
    crc = 0xFFFFFFFF;
    for (char byte : binaryMessage) {
        crc = (crc >> 8) ^ crcTable[(crc ^ byte) & 0xFF];
    }

    return crc ^ 0xFFFFFFFF;
}

int main() {
    std::cout << "CRC-32 Checksum Receiver" << std::endl;
    std::cout << "------------------------" << std::endl;

    std::string fullMessage;
    std::cout << "Enter the binary message (with checksum): ";
    std::cin >> fullMessage;

    std::string binaryMessage = fullMessage.substr(0, fullMessage.length() - 32);
    std::string receivedChecksum = fullMessage.substr(fullMessage.length() - 32);

    uint32_t crc32Checksum = calculateCRC32(binaryMessage);

    std::bitset<32> computedChecksumBits(crc32Checksum);
    std::string computedChecksumBinary = computedChecksumBits.to_string();

    if (computedChecksumBinary == receivedChecksum) {
        std::cout << "Message is VALID." << std::endl;
    } else {
        std::cout << "Message is CORRUPTED." << std::endl;
    }

    return 0;
}

File: CRC-32/Python/receiver.py
def calculateChecksum(binaryMessage):
    crc = 0xFFFFFFFF
    table = [0] * 256
    polynomial = 0xEDB88320

    # Generate CRC-32 table
    for i in range(256):
        crc = i
        for j in range(8):
            if crc & 1:
                crc = (crc >> 1) ^ polynomial
            else:
                crc >>= 1
        table[i] = crc

    # Calculate CRC-32
    crc = 0xFFFFFFFF
    for byte in binaryMessage:
        crc = (crc >> 8) ^ table[(crc ^ ord(byte)) & 0xFF]

    return crc ^ 0xFFFFFFFF

if __name__ == "__main__":
    print("----- CRC-32 Checksum Decoder -----")

    fullMessage = input("Enter the binary message (with checksum): ")

    binaryMessage = fullMessage[:len(fullMessage) - 32]
    receivedChecksum = fullMessage[-32:]

    checksum = calculateChecksum(binaryMessage)

    checksumBinary = bin(checksum & 0xFFFFFFFF)[2:].zfill(32)

    if checksumBinary == receivedChecksum:
        print("Message is VALID.")
    else:
        print("Message is CORRUPTED.")
